File: java-workspace2/src/분류안함/설탕배달.java
package 분류안함;

import java.util.Arrays;
import java.util.Scanner;

public class 설탕배달 {
	static Scanner scanner = new Scanner(System.in);

	public static void main(String[] args) {
		int target = scanner.nextInt();
		int[] units = new int[]{3, 5};
		int[] dp = new int[target + 1];
		int INF = Integer.MAX_VALUE;
		Arrays.fill(dp, INF);
		dp[0] = 0;
		for (int unit : units) {
			for (int weight = unit; weight <= target; weight++) {
				if (dp[weight - unit] != INF) {
					dp[weight] = Math.min(dp[weight], dp[weight - unit] + 1);
				}
			}
		}
		System.out.println(dp[target] == INF ? -1 : dp[target]);
	}
}

File: python-workspace/백준/다이나믹프로그래밍/설탕배달.py
import sys

units = [5, 3]
target = int(sys.stdin.readline())
INF = 0xffffff
dp = [INF for _ in range(target + 1)]
dp[0] = 0
for unit in units:
    for num in range(unit, target + 1):
        if dp[num - unit] != INF:
            dp[num] = min(dp[num], dp[num - unit] + 1)
print(dp[target] if dp[target] != INF else -1)
